core/arch/arm/plat-hikey/spi_test.c

```diff
@@ -28,73 +28,246 @@
- * Mezzanine card [1], which can be connected to either a sliding
- * rheostat [2] or photoresistor [3].
- *
- * [1] http://linksprite.com/wiki/index.php5?title=Linker_Mezzanine_card_for_96board
- * [2] http://learn.linksprite.com/96-board/sliding-rheostat
- * [3] http://learn.linksprite.com/96-board/photoresistor
- */
-void spi_test(void)
+#define PL022_STAT	0x00C
+#define PL022_STAT_BSY	SHIFT_U32(1, 4)
+
+static void spi_cs_callback(enum gpio_level value)
 {
+	static bool is_init = false;
 	struct pl061_data platform_pl061_data;
-	struct pl022_data platform_pl022_data;
 	vaddr_t gpio6_base = nsec_periph_base(GPIO6_BASE);
 	vaddr_t spi_base = nsec_periph_base(SPI_BASE);
+
+	if (!is_init) {
+		pl061_init(&platform_pl061_data);
+		pl061_register(gpio6_base, 6);
+		pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
+		platform_pl061_data.chip.ops->set_interrupt(GPIO6_2,
+						GPIO_INTERRUPT_DISABLE);
+		platform_pl061_data.chip.ops->set_direction(GPIO6_2,
+						GPIO_DIR_OUT);
+		is_init = true;
+	}
+
+	if (read8(spi_base + SSPSR) & SSPSR_BSY)
+		DMSG("pl022 busy - do NOT set CS!");
+	while (read8(pd->base + SSPSR) & SSPSR_BSY)
+		;
+	DMSG("pl022 done - set CS!\n");
+
+	platform_pl061_data.chip.ops->set_value(GPIO6_2, value);
+}
+
+static void spi_set_cs_mux(uint32_t val)
+{
+	uint32_t data;
+	vaddr_t pmx0_base = nsec_periph_base(PMX0_BASE);
+
+	if (val == PINMUX_SPI) {
+		DMSG("configure gpio6 pin2 as SPI\n");
+		write32(PINMUX_SPI, pmx0_base + PMX0_IOMG106);
+	} else {
+		DMSG("configure gpio6 pin2 as GPIO\n");
+		write32(PINMUX_GPIO, pmx0_base + PMX0_IOMG106);
+	}
+
+	data = read32(pmx0_base + PMX0_IOMG106);
+	if (data)
+		DMSG("gpio6 pin2 is SPI\n");
+	else
+		DMSG("gpio6 pin2 is GPIO\n");
+}
+
+static void spi_test_with_manual_cs_control(void)
+{
+	struct pl022_data platform_pl022_data;
+	vaddr_t spi_base = nsec_periph_base(SPI_BASE);
 	uint8_t tx[3] = {0x01, 0x80, 0x00};
 	uint8_t rx[3] = {0};
-	size_t i, j, num_rxpkts, len = 3;
+	size_t i, j, len = 3;
+
+	spi_set_cs_mux(PINMUX_GPIO);
+
+	DMSG("set CS callback\n");
+	platform_pl022_data.cs_data.cs_cb = NULL;
+	platform_pl022_data.b_cs_callback = true;
+
+	DMSG("spi_base: 0x%" PRIxVA "\n", spi_base);
+	DMSG("configure SPI\n");
+	platform_pl022_data.base = spi_base;
+	platform_pl022_data.clk_hz = SPI_CLK_HZ;
+	platform_pl022_data.speed_hz = SPI_10_KHZ;
+	platform_pl022_data.mode = SPI_MODE0;
+	platform_pl022_data.data_size_bits = 8;
+	platform_pl022_data.loopback = true;
+
+	pl022_init(&platform_pl022_data);
+	platform_pl022_data.chip.ops->configure(&platform_pl022_data.chip);
+	platform_pl022_data.chip.ops->start(&platform_pl022_data.chip);
+
+	/*
+	 * Pulse CS only once for the whole transmission.
+	 * This is the scheme used by the pl022 driver.
+	 */
+	spi_cs_callback(GPIO_LEVEL_HIGH);
+	tee_time_busy_wait(2);
+	spi_cs_callback(GPIO_LEVEL_LOW);
+	for (j = 0; j < 10; j++) {
+		DMSG("SPI test loop: %zu\n", j);
+		platform_pl022_data.chip.ops->txrx8(&platform_pl022_data.chip,
+						tx, rx, len);
```

Comment:
Should test the return value. Here and at several other places.
Same regarding `txrx16` handler.